### PR TITLE
Common templating tools

### LINKED ISSUE #69: Common templating tools
**Is your feature request related to a problem? Please describe.**
Many of the drivers support the use of template syntax to inject resource property, system property and deployment location property values into infrastructure templates/lifecycle scripts. 

The template requirements are often the same but the solutions differ slightly in each driver (e.g. reference names such as dl_properties in one driver, deploymentLocation in another)

**Describe the solution you'd like**
To encourage consistent syntax Ignition could provide:

- Common library for rendering templates
- Common library for building the collection of properties which may be referenced in a template (render props)

### PR DESCRIPTION
**Pull request by dvaccarosenna**
Closes #69

### CODE REVIEW

File: docs/user-guide/framework/bootstrap-components/templating.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+Although optional, the Services listed below are enabled by default when using the `ignition.boot.api.build_vim_driver` or `ignition.boot.api.build_lifecycle_driver` method. 
+
+## Services
+
+The following services are auto-configured when enabled:
+
+| Name                     | Capability         | Required Capabilities             | Bootstrap Enable/Disable flag       | Description                                                                                                                                     |
+| ------------------------ | ------------------ | --------------------------------- | ----------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- |
+| Jinja2TemplatingService | JobQueTemplatingCapabilityueCapability | - | bootstrap.templating.service_enabled | Handles rendering templates |
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/templating.md:33]: JobQueTemplatingCapabilityueCapability - typo?

File: ignition/service/templating.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Implementation of the TemplatingCapability based on Jinja2
+    """
+
+    def __build(self, content):
+        return JinjaTemplate(content)
+
+    def render(self, content, context):
+        template = self.__build(content)
+        return template.render(context)
```
Comment (sglover) [inline ignition/service/templating.py:33]: Do you need to assign to a variable first - can't you just do it on one line?

File: ignition/templating/resource_context_builder.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        Add extra property. If the property is already present the existing value will be replaced by the incoming value
+
+        Args:
+            key (str): name of the property
+            value: value to assign to the property
+
+        Returns:
+            this builder
+        """
+        self.result[key] = value
```
Comment (sglover) [inline ignition/templating/resource_context_builder.py:84]: I guess this needs the same check for reserved word as above?

File: tests/unit/templating/test_resource_context_builder.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                'sysB': 'B'
+            }, 
+            'deploymentLocationInst': {
+                'name': 'Test',
+                'type': 'Kubernetes',
+                'properties': {
+                    'dlPropA': 'A DL Prop'
+                }
+            }
+        })
```
Comment (sglover) [inline tests/unit/templating/test_resource_context_builder.py:251]: Can't see any tests where you try to set property or properties with key "systemProperties" and "deploymentLocationInst" (which are covered in the code) - it's possible I've missed these tests though!

File: docs/user-guide/framework/bootstrap-components/templating.md (hunk 1 above)
Comment (dvaccarosenna) [reply docs/user-guide/framework/bootstrap-components/templating.md:33]: copy and paste error :man_facepalming:

File: tests/unit/templating/test_resource_context_builder.py (hunk 4 above)
Comment (dvaccarosenna) [reply tests/unit/templating/test_resource_context_builder.py:251]: you're right, they are missing. I will add them.

**Pull request merged**